Error starting LSP Qute

Since the last plugin update I get the LSP error provided below on IntelliJ start and then the IDE has some LSP features missing as a whole. I have upgraded the IntelliJ version, did not help. You can find the configuration listed below. Any idea what I am doing wrong?

IntelliJ: Build #IC-231.9011.34, built on May 16, 2023
Runtime version: 17.0.6+10-b829.9 aarch64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Quarkus Tools: 1.19.0.388
RedHat Telemetry: 1.0.0.44
OS: macOS Venture 13.4 / M1 Chip

The error:

```
Error while starting language server 'qute'

java.util.concurrent.CompletionException: org.eclipse.lsp4j.jsonrpc.JsonRpcException: java.io.IOException: Stream closed
	at java.base/java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:368)
	at java.base/java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:377)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1152)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1773)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: org.eclipse.lsp4j.jsonrpc.JsonRpcException: java.io.IOException: Stream closed
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageConsumer.consume(StreamMessageConsumer.java:72)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.LanguageServerWrapper.lambda$start$1(LanguageServerWrapper.java:258)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.request(RemoteEndpoint.java:161)
	at org.eclipse.lsp4j.jsonrpc.services.EndpointProxy.invoke(EndpointProxy.java:91)
	at jdk.proxy6/jdk.proxy6.$Proxy126.initialize(Unknown Source)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.LanguageServerWrapper.initServer(LanguageServerWrapper.java:365)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.LanguageServerWrapper.lambda$start$4(LanguageServerWrapper.java:283)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
	... 8 more
Caused by: java.io.IOException: Stream closed
	at java.base/java.lang.ProcessBuilder$NullOutputStream.write(ProcessBuilder.java:445)
	at java.base/java.io.OutputStream.write(OutputStream.java:162)
	at java.base/java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:81)
	at java.base/java.io.BufferedOutputStream.flush(BufferedOutputStream.java:142)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageConsumer.consume(StreamMessageConsumer.java:69)
	... 15 more
```

- The issue is permanent in my case, 1.19.0 breaks LSP features in the IDE on my system.

- I had already tried ensuring that no additional processes are running. I also restarted the whole system and only started the IDE but does not help. `jsp` reports only itself, the IDE process and RemoteMavenServer36. 

- I installed the latest CI build available just a few minutes ago (https://github.com/redhat-developer/intellij-quarkus/suites/13457228985/artifacts/738401212) and the issue still persists.

- I then installed Quarkus Tools 1.17.1 and it works for me. The features that I need for my development are there, though it throws another exception, which I have been ignoring all along. 

- I could check when the functionality broke, but I cannot find any intermediate build artefacts in the CI builds.

- I am only supplying the exception of 1.17.1 in case it is somehow connected to the current issue in 1.19.0, but it looks like a version mismatch with IntelliJ.

```
Internal error: java.lang.NoClassDefFoundError: org/jetbrains/plugins/gradle/util/GradleConstants


java.util.concurrent.CompletionException: java.lang.NoClassDefFoundError: org/jetbrains/plugins/gradle/util/GradleConstants
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:708)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.IndexAwareLanguageClient.runTask(IndexAwareLanguageClient.java:101)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.IndexAwareLanguageClient$1.run(IndexAwareLanguageClient.java:76)
	at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:429)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:114)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressSynchronously$9(CoreProgressManager.java:513)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$new$0(ProgressRunner.java:84)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:252)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:186)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:604)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:679)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:635)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:603)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:173)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:252)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.NoClassDefFoundError: org/jetbrains/plugins/gradle/util/GradleConstants
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.providers.GradleProjectLabelProvider.isGradleProject(GradleProjectLabelProvider.java:45)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.internal.core.providers.GradleProjectLabelProvider.getProjectLabels(GradleProjectLabelProvider.java:38)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.ProjectLabelManager.getProjectLabels(ProjectLabelManager.java:122)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.ProjectLabelManager.getProjectLabelInfo(ProjectLabelManager.java:77)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.ProjectLabelManager.lambda$getProjectLabelInfo$1(ProjectLabelManager.java:102)
	at com.intellij.openapi.project.DumbService.lambda$runReadActionInSmartMode$0(DumbService.java:112)
	at com.intellij.openapi.project.DumbService.lambda$runReadActionInSmartMode$1(DumbService.java:156)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:891)
	at com.intellij.openapi.application.ReadAction.compute(ReadAction.java:69)
	at com.intellij.openapi.project.DumbService.runReadActionInSmartMode(DumbService.java:149)
	at com.intellij.openapi.project.DumbService.runReadActionInSmartMode(DumbService.java:112)
	at com.redhat.devtools.intellij.lsp4mp4ij.psi.core.ProjectLabelManager.getProjectLabelInfo(ProjectLabelManager.java:102)
	at com.redhat.devtools.intellij.quarkus.lsp.QuarkusLanguageClient.lambda$getJavaProjectlabels$6(QuarkusLanguageClient.java:132)
	at com.redhat.devtools.intellij.quarkus.lsp4ij.IndexAwareLanguageClient.runTask(IndexAwareLanguageClient.java:97)
	... 22 more
Caused by: java.lang.ClassNotFoundException: org.jetbrains.plugins.gradle.util.GradleConstants PluginClassLoader(plugin=PluginDescriptor(name=Quarkus Tools, id=com.redhat.devtools.intellij.quarkus, descriptorPath=plugin.xml, path=~/Library/Application Support/JetBrains/IdeaIC2023.1/plugins/Quarkus Tools, version=1.17.1, package=null, isBundled=false), packagePrefix=null, instanceId=30, state=active)
	... 36 more
```

Thank you for the PR and your help, I will check that. 

Starting the IDE from a terminal (macOS Terminal) works, no exceptions are thrown on start and the LSP functionality is available. I still get a different exception in the terminal, listed below, but the IDE seems fine.  

I looked at the referenced changes, and I am not completely sure what the difference is. I am using SDKMAN! for the Java environments, as listed above the IDE uses its own v17 to start. 

SDKMAN! configures the environment vars for 11.0.17-zulu, under "/Users/tsept/.sdkman/candidates/java/current/". I noticed that the projects were somehow set up for 11.0.16-zulu. I changed them to the environment version (11.0.17-zulu), but this did not resolve the issue when I start IntelliJ normally, without using the terminal.

The console exception is:

```
Jun 08, 2023 3:11:23 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.NullPointerException
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:645)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:212)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:193)
	at org.eclipse.lsp4mp.ls.commons.ParentProcessWatcher.lambda$apply$0(ParentProcessWatcher.java:148)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	at org.eclipse.lsp4mp.ls.java.JavaFileTextDocumentService.lambda$completion$3(JavaFileTextDocumentService.java:173)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	... 13 more
```

--- a/src/main/java/com/redhat/devtools/intellij/lsp4ij/server/JavaProcessCommandBuilder.java
+++ b/src/main/java/com/redhat/devtools/intellij/lsp4ij/server/JavaProcessCommandBuilder.java
@@ -73,7 +73,7 @@ public class JavaProcessCommandBuilder {
 
     public List<String> create() {
         List<String> commands = new ArrayList<>();
-        commands.add(computeJavaPath());
+        commands.add(javaPath);
         if (debugPort != null && !debugPort.isEmpty()) {
             String suspend = debugSuspend ? "y" : "n";
             commands.add("-agentlib:jdwp=transport=dt_socket,server=y,suspend=" + suspend + ",address=" + debugPort);
@@ -90,14 +90,7 @@ public class JavaProcessCommandBuilder {
     }
 
     private static String computeJavaPath() {
-        String javaPath = "java";
-        boolean existsInPath = Stream.of(System.getenv("PATH").split(Pattern.quote(File.pathSeparator))).map(Paths::get)
-                .anyMatch(path -> Files.exists(path.resolve("java")));
-        if (!existsInPath) {
-            File f = new File(System.getProperty("java.home"),
-                    "bin/java" + (OS.current() == OS.WINDOWS ? ".exe" : ""));
-            javaPath = f.getAbsolutePath();
-        }
-        return javaPath;
+        return new File(System.getProperty("java.home"),
+                "bin/java" + (OS.current() == OS.WINDOWS ? ".exe" : "")).getAbsolutePath();
     }
 }

